### ISSUE TITLE
Linux Setup

### ISSUE THREAD

**Issue by JorjMcKie** [rk700/PyMuPDF#35]
@rk700 - trying to set-up my own Linux version, but I am a newbie there. Therefore a question: how do I generate the required MuPDF third-party libraries? That's the only thing I am missing right now.

**Comment by rk700** [comment]
I think this depends on the Linux distribution since different distributions may use different packaging systems. 

Take mine as an example: I'm using archlinux, and here's the detail info of the mupdf package on archlinux: https://www.archlinux.org/packages/community/x86_64/mupdf/, which shows that mupdf depends on several other libraries, including freetype2, libjpeg, etc. These libraries can be installed from official repo.

It seems that @ousia is using fedora, and I think some similar mechanism would ensure that the 3rd party libraries are available before installing mupdf.

So which Linux distribution are you using? If you can find mupdf in the official repo, then usually the 3rd party libraries can also be found there. It would be fine to just install them with the package management tool(apt-get, yum, pacman, etc.). However if the libraries cannot be found in the repo, then maybe you  have to build the 3rd party libraries from source.

**Comment by JorjMcKie** [comment]
I am using Ubuntu. I just naively tried to execute make from the root directory of the MuPDF download. As far as the command line modules are concerned (mutool, etc.) everything is there on the 1.8 release level. But now, for generating fitz, I need all so libraries ...

**Comment by rk700** [comment]
It seems that Ubuntu splits mupdf into the viewer and development headers/libs:
http://packages.ubuntu.com/trusty/mupdf
http://packages.ubuntu.com/trusty/libmupdf-dev

So I think you can just install mupdf and libmupdf-dev from the repo(apt-get install) and the libs would be installed. I'll try that in a virtual machine and let you know the result.

**Comment by rk700** [comment]
I've tried on Ubuntu 14.04. Almost done but still some annoying issues.
* Mupdf in the official repo is 1.3, which is TOO low. I have to compile it from source. Also I have to add '-fPIC' option to the CFLAGS in mupdf's Makefile for symbol resolution.
* Then I need to install some dev packages for PyMuPDF(I already installed some when building mupdf). To name a few, python-dev, libjpeg-turbo8-dev, etc. Here's the place I used for looking up which package contains the library I need: http://packages.ubuntu.com/search?mode=exactfilename&suite=trusty&section=all&arch=any&keywords=libjbig2dec.so&searchon=contents
* All the 3rd party libraries can be found in the repo except libopenjp2.so. This one is not available in Ubuntu until very recently release, http://packages.ubuntu.com/xenial/libopenjp2-7. I just downloaded the source and again built it
* Finally I realize that the swig on Ubuntu made several really annoying changes to the output .py and .c, which corrupts the library. I've no way but to change the setup.py and use the .py and .c I generated before. I'll commit the change later on.

To sum up, it takes a little extra work to build PyMuPDF on Ubuntu because the package versions are too low. Maybe I should add the detail instructions since I believe that most Ubuntu users are still using old ones like 14.04 or before.

**Comment by JorjMcKie** [comment]
Uff! All this effort caused by a small question.
It seems to me, that providing the compiled libraries as we had it before was not such a bad idea. Image a normal Ubuntu (e.g. LinuxMint) user ... he will probably never use our tool in face of all that hassle!
I suggest to include the Linux libs in the new Optional Material repository just for those people who do not have a Linux version where everything works smoothely as it apparently does on Fedora.

**Comment by rk700** [comment]
I think it's better not to provide the libraries for 3 reasons:

1. The library dependencies can be very complex. MuPDF and PyMuPDF depend on several 3rd party libraries, and these libraries also depend on other base libraries too.We would have to create a large amount of libraries since different Linux distributions may have different versions of the libraries.

2. Linux users would like to compile from source instead of using binaries from someone they don't trust. This is the why Fedora package maintainer asked to remove the libraries we made before. And I believe that it's the same for all other Linux distributions.

3. PyMuPDF has an implicit dependency on MuPDF. If the user would like to use PyMuPDF 1.8, then he should first make sure that MuPDF 1.8 is available. And as long as MuPDF 1.8 is available, then all the 3rd party libraries should already be somewhere in the system. So, for distributions where MuPDF 1.8 is in the official repo(like archlinux I'm using), the 3rd party libraries can also be found in the repo; for distributions where MuPDF 1.8 is not in the official repo(like ubuntu), the 3rd party libraries are built when compiling MuPDF from the source.

I didn't take a look at mupdf directory last week, and I just realize that the 3rd party libraries, like libopenjpeg.a, libjbig2dec.a, etc., can be found in the `build` directory once make finished. So ubuntu users would only need to add the library path to setup.py for building PyMuPDF without dealing with dependencies. I'll update the script later with instructions.

**Comment by rk700** [comment]
With the README updated, it should be clear now for Ubuntu users.

**Comment by JorjMcKie** [comment]
ok, thanks. I will try it out as a Ubuntu novice ... :-)
What happened to the problem you mentioned about SWIG output?

**Comment by rk700** [comment]
SWIG on my Archlinux is of version 3, while on Ubuntu it's of version 2. I guess this might result in the unexpected output.

**Comment by JorjMcKie** [comment]
Should we not leave out the SWIG step from setup alltogether and just take the pre-generated fitz_wrap.c as source?
Otherweise we would have to announce SWIG as a pre-requisite for PyMuPDF installation and also inttroduce another difference to the Windows installation.

**Comment by rk700** [comment]
Yes I've updated the `setup.py` and use `fitz_wrap.c` as source instead of `fitz.i`. So now SWIG is not required if a user want to build PyMuPDF, but as developers we should keep it since we need SWIG to generate the .py and .c files after editing `fitz.i`
